### PR TITLE
[ios] introduce weak_nsobject

### PR DESCRIPTION
**Pull request by cyanglaz**
Introduce weak_nsobject from chromium. 

There are some usages of weak_ptr wrapping Objective-C ids, weak_ptr is not really designed for ids and such usages are blocking the arc migration. 

This PR mostly copies the weak_nsobject from chromium, at the same hash that we copied the ARC/MRC compatible scoped_nsobject: https://chromium.googlesource.com/chromium/src/+/fd625125b8a6a3aceaf09993a5f74cfe5368b17f

To match how we used weak_ptr for those ids, I made some changes to the weak_nsobject:
- WeakNSObjects needs to be generated by a WeakNSObjectFactory. The WeakNSObjectFactory is owned by the objc class and acts as the generator of the WeakNSObjects. All the WeakNSObjects' derefing thread should be the same of the WeakNSObjectFactory's creation thread.
- chromuim's WeakNSObjects can be detached from the thread and re-attached to a new thread. To match our weak_ptr behavior, I changed WeakNSObjects to be only accessed from a single thread, the same as weak_ptr

This PR also moves the FlutterEngine to use WeakNSObject and updated related classes.

part of https://github.com/flutter/flutter/issues/137801

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by cyanglaz** [comment]
I'm going to migrate all the weakPtr<id> to weakNSObject in this PR, convert to draft until that;s done

File: fml/platform/darwin/weak_nsobject.h [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    CheckThreadSafety();
+    object_ = nil;
+  }
+
+ private:
+  friend fml::RefCountedThreadSafe<WeakContainer>;
+  ~WeakContainer() {}
+
+  std::shared_ptr<debug::DebugThreadChecker> checker_;
+  id object_;
```
Comment (stuartmorgan) [inline fml/platform/darwin/weak_nsobject.h:89]: This doesn't look ARC-safe to me; this is a strong reference under ARC unless I'm misremembering the semantics of `id` in ARC. I was confused by this so I looked at the history, and although:

> This PR mostly copies the weak_nsobject from chromium, at the same hash that we copied the ARC/MRC compatible scoped_nsobject: https://chromium.googlesource.com/chromium/src/+/fd625125b8a6a3aceaf09993a5f74cfe5368b17f

that doesn't appear to have been an ARC-safe version. See https://codereview.chromium.org/2943263003 which landed much later.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: fml/platform/darwin/weak_nsobject.h (hunk 1 above)
Comment (cyanglaz) [reply fml/platform/darwin/weak_nsobject.h]: Thanks for catching this. I have adopted the change and updated the NSWeakObjectFactory accordingly.
I have also added a new unittests to test the NSWeakObject under ARC.
Also migrated a XCTest file that contains NSWeakObject to ARC.

File: fml/platform/darwin/weak_nsobject.mm [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  FML_DCHECK(container.get());
+  self = [super init];
+  if (self) {
+    container_ = container;
+  }
+  return self;
+}
+
+- (void)dealloc {
+  self.container->nullify();
```
Comment (stuartmorgan) [inline fml/platform/darwin/weak_nsobject.mm:65]: Nit: `container_`; `dealloc` shouldn't use `self`.

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Pull request merged**